src/oss/langchain-messages.mdx

```diff
@@ -2,14 +2,488 @@
+    for tool_call in tool_calls:
+        result = execute_tool(tool_call["name"], tool_call["args"])
+        tool_msg = ToolMessage(
+            content=str(result),
+            tool_call_id=tool_call["id"]
+        )
+        messages.append(tool_msg)
+    return messages
+
+# Use in conversation loop
+messages = [HumanMessage("Calculate 25 * 4 and get weather for NYC")]
+response = llm_with_tools.invoke(messages)
+
+if response.tool_calls:
+    messages.append(response)  # Add AI response with tool calls
+    messages = handle_tool_calls(messages, response.tool_calls)
+    final_response = model.invoke(messages)  # Get final answer
+```
+
 ## Content Blocks
+
+Content blocks enable **provider-agnostic** multimodal interactions. Instead of handling different image formats, audio types, and document structures across providers, content blocks provide a unified interface that works everywhere (supported).
+
+TODO link out to guide
+
+### Why Content Blocks Matter
+
+TODO this needs to be a gif that shows the massive differences at a glance
+
+```python
+# ❌ Provider-specific multimodal handling
+openai_message = {
+    "role": "user", 
+    "content": [
+        {"type": "text", "text": "What's in this image?"}, 
+        {"type": "image_url", "image_url": {"url": "..."}}  # OpenAI format
+    ]
+}
+
+anthropic_message = {
+    "role": "user",
+    "content": [
+        {"type": "text", "text": "What's in this image?"},
+        {"type": "image", "source": {"type": "url", "media_type": "image/jpeg", "url": "..."}}  # Anthropic format
+    ]
+}
+
+# ✅ Universal content blocks
+universal_message = HumanMessage([
+    {"type": "text", "text": "What's in this image?"},
+    {"type": "image", "source_type": "url", "url": "..."}  # Works with any provider
+])
+```
+
+### Standard Content Types
+
+TODO switcher to show different content types
+
+```python
+# Text content
+{"type": "text", "text": "Hello world"}
+
+# Images (URL or base64)  
+{"type": "image", "source_type": "url", "url": "https://..."}
+{"type": "image", "source_type": "base64", "data": "...", "mime_type": "image/jpeg"}
+
+# Documents
+{"type": "file", "source_type": "base64", "data": "...", "mime_type": "application/pdf"}
+
+# Audio
+{"type": "audio", "source_type": "base64", "data": "...", "mime_type": "audio/wav"}
+```
+
+### Multimodal Message Example
+
+```python
+# Combine multiple content types in one message
+multimodal_message = HumanMessage([
+    {"type": "text", "text": "Compare this image and document:"},
+    {"type": "image", "source_type": "url", "url": "chart.jpg"},
+    {"type": "file", "source_type": "base64", "data": pdf_data, "mime_type": "application/pdf"},
+    {"type": "text", "text": "Which data source is more reliable?"}
+])
+
+# Works with any compatible model
+response = model.invoke([multimodal_message])
+```
+
+### Rich Model Responses
+
+With `output_version="v1"`, models return structured content blocks:
```

Comment:
needs better explanation. all messages will have .content_blcoks right? so this doesnt motivate it fully